OpenRA.Mods.Common/Traits/PaletteEffects/MenuPostProcessEffect.cs

```diff
@@ -19,61 +19,82 @@ namespace OpenRA.Mods.Common.Traits
 	[Desc("Fades the world from/to black at the start/end of the game, and can (optionally) desaturate the world")]
 	public class MenuPostProcessEffectInfo : TraitInfo
 	{
+		public readonly int FadeInLength = 10;
+
 		[Desc("Time (in ticks) to fade between states")]
 		public readonly int FadeLength = 10;
 
 		[Desc("Effect style to fade to during gameplay. Accepts values of None or Desaturated.")]
 		public readonly MenuPostProcessEffect.EffectType Effect = MenuPostProcessEffect.EffectType.None;
 
+		[Desc("Effect style to fade from when starting the game. Accepts values of None, Black or Desaturated.")]
+		public readonly MenuPostProcessEffect.EffectType GameStartEffect = MenuPostProcessEffect.EffectType.Black;
+
+		[Desc("Effect style to fade to when exiting the game. Accepts values of None, Black or Desaturated.")]
+		public readonly MenuPostProcessEffect.EffectType GameExitEffect = MenuPostProcessEffect.EffectType.Black;
+
 		[Desc("Effect style to fade to when opening the in-game menu. Accepts values of None, Black or Desaturated.")]
 		public readonly MenuPostProcessEffect.EffectType MenuEffect = MenuPostProcessEffect.EffectType.None;
 
 		public override object Create(ActorInitializer init) { return new MenuPostProcessEffect(this); }
 	}
 
-	public class MenuPostProcessEffect : RenderPostProcessPassBase, IWorldLoaded, INotifyGameLoaded
+	public class MenuPostProcessEffect : RenderPostProcessPassBase, INotifyGameLoaded, IPostWorldLoaded
 	{
 		public enum EffectType { None, Black, Desaturated }
 		public readonly MenuPostProcessEffectInfo Info;
 
-		EffectType from = EffectType.Black;
-		EffectType to = EffectType.Black;
+		EffectType from;
+		EffectType to;
 
 		long startTime;
 		long endTime;
+		float blend;
 
 		public MenuPostProcessEffect(MenuPostProcessEffectInfo info)
 			: base("menufade", PostProcessPassType.AfterShroud)
 		{
 			Info = info;
+			to = info.GameStartEffect;
 		}
 
-		public void Fade(EffectType type)
+		public void Fade(EffectType type, int length = -1)
 		{
+			if (length < 0)
+				length = Info.FadeLength;
+
 			startTime = Game.RunTime;
-			endTime = startTime + Ui.Timestep * Info.FadeLength;
+			endTime = startTime + Ui.Timestep * length;
 
 			from = to;
 			to = type;
 		}
 
-		protected override bool Enabled => to != EffectType.None || endTime != 0;
-		protected override void PrepareRender(WorldRenderer wr, IShader shader)
+		float SetBlend()
 		{
-			var blend = (endTime - Game.RunTime) * 1f / (endTime - startTime);
+			if (startTime == 0)
+				return 0;
+
+			blend = (endTime - Game.RunTime) * 1f / (endTime - startTime);
 			if (blend < 0)
 				blend = startTime = endTime = 0;
 
+			return blend;
+		}
+
+		protected override bool Enabled => SetBlend() != 0 || to != EffectType.None;
```

Comment:
Having the rendering logic driven by a side-effect of checking whether the effect is active isn't great.

This could be checking `Game.RunTime < endTime || to != EffectType.None` allowing the blend calculation to stay within `PrepareRender`.

Reply:
adjusted